insights/client/crypto.py

```diff
@@ -0,0 +1,122 @@
+import os.path
+import logging
+import shutil
+import tempfile
+import shlex
+import subprocess
+
+from .constants import InsightsConstants as constants
+
+logger = logging.getLogger(__file__)
+
+
+def _call_gpg(home, command):
+    """
+    Run /usr/bin/gpg.
+
+    Parameters:
+        home (str): Path to GPG's home directory.
+        command (str): Actual arguments passed to the GPG binary.
+
+    Returns (str, str, int):
+        Standard output, standard error and the return code.
+    """
+    command = (
+        "/usr/bin/gpg --homedir {home} {command}"
+    ).format(home=home, command=command)
+    process = subprocess.Popen(shlex.split(command), stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    raw_stdout, raw_stderr = process.communicate()
+    stdout, stderr = raw_stdout.decode("utf-8"), raw_stderr.decode("utf-8")
+
+    logger.debug(command)
+    logger.debug("status code {code}".format(code=process.returncode))
+
+    if process.returncode > 0 and stdout:
+        logger.debug("stdout is not empty:\n%s", stdout.strip())
+    if process.returncode > 0 and stderr:
+        logger.debug("stderr is not empty:\n%s", stderr.strip())
+
+    return stdout, stderr, process.returncode
+
+
+def verify_gpg_signed_file(file, signature):
+    """
+    Verify a file that was signed using GPG.
+
+    Parameters:
+        file (str): A path to the signed file.
+        signature (str): A path to the detached signature.
+
+    Returns (dict):
+        ok (str): Whether the signature holds.
+        stdout (str): Standard output of GPG.
+        stderr (str): Standard error of GPG.
+        code (int): Return code of GPG.
+    """
+    if not os.path.isfile(file):
+        message = "cannot verify signature of '{file}', file does not exist".format(file=file)
+        logger.debug(message)
+        return {
+            'gpg': False,
+            'stdout': '',
+            'stderr': '',
+            'rc': 1,
+        }
```

Comment:
* Unless it's a standard throughout the insights code base not to use exceptions for errors, it would be better to raise an exception for errors like this.
* I only see one caller that is using the stdout, stderr, and rc. (in `client/__init__.py:InsightsClient.verify()`  I would check whether it's okay to only use those values for error cases (where they can be attributes of the exception) and then change the API of this function to return a simple boolean True (gpg ran and the signature was valid) |False (gpg ran and the signature was invalid).
 * In fact, a quick look at `client/__init__.py` seems to show a bug: https://github.com/RedHatInsights/insights-core/pull/3930/files#diff-5ef20b67881931ced25261142697d774de208b9c1c2571da00294e3a84a76038R263  is expecting a boolean True or False already so `verify()` in this if statement: `if (egg_paths and self.verify(egg_paths['core'])['gpg']):` will always be truthy in both the current code and your revision.